### PR TITLE
test: experiment list page models and sample test [INFENG-451]

### PR DESCRIPTION
**Pull request by JComins000**
## Ticket

[INFENG-451]

## Description

I didn't test everything, but I created a lot of extra scaffolding on top of the base requirement for INFENG-451. For the scope of that ticket, all we care about is being able to read and click on the `DataGrid` in `F_ExperimentList`, which is certainly possible now!

Mine looked like this and I was able to click the select column and read values from the table
image

I couldn't help but rearrange some logic in BaseComponent.ts again. All the components are the same but there's a new interface that things implement from and I renamed `ParentTypes`.

URLs can be regex now as well. I needed to change that for the projectdetails page. I added some guardrails too.

Also, I was lazy about creating a `hew/Pagination` model last time, but this time I needed to reuse it. It's in the proper place now.

Lastly, this probably deserves some scrutiny. If you're going to criticize it, I would recommend checking out the branch and trying to make changes directly. The reason it's so complicated is that one of the attributes of `Row` is a reference back to it's parent table `Table<Row, HeadRow>`, and so then Row needs to look like `Row<Row, HeadRow>`. Even though Row doesn't use HeadRow directly, it still needs that knowledge to properly type it's reference to the parent table.
```
type RowClass<RowType extends Row<RowType, HeadRowType>, HeadRowType extends HeadRow> = new (
  args: RowArgs<RowType, HeadRowType>,
) => RowType;
```

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[INFENG-451]: https://hpe-aiatscale.atlassian.net/browse/INFENG-451?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: webui/react/src/pages/ExperimentList.tsx [hunk 1]
```diff
@@ -962,6 +962,7 @@ const ExperimentList: React.FC<Props> = ({ project }) => {
           label: action,
           value: action,
         }))}
+        data-test-component="experiementList-tableBatch"
```
Comment (JComins000) [inline webui/react/src/pages/ExperimentList.tsx:965]: im gonna remove this

File: webui/react/src/e2e/models/hew/DataGrid.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    const matches = style.match(/height: (\d+)px;/);
+    if (matches === null) {
+      throw new Error("Couldn't find height in style attribute.");
+    }
+    this.#columnheight = +matches[1];
+    return this.columnHeight;
+  }
+
+  /**
+   * Returns a row from an index. Start counting at 1.
```
Comment (ashtonG) [inline webui/react/src/e2e/models/hew/DataGrid.ts:96]: why are we doing 1-based indexing?

Comment (JComins000) [reply webui/react/src/e2e/models/hew/DataGrid.ts:96]: Datagrid starts counting at 1 for it's head row. And row 2 is the first row of the body.

Comment (JComins000) [reply webui/react/src/e2e/models/hew/DataGrid.ts:96]: Actually, you're right. I can just use n+2

File: webui/react/src/e2e/models/hew/DataGrid.ts [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+      }),
+    );
+    return this.#columnDefs;
+  }
+
+  /**
+   * Clicks the head row's select button
+   */
+  async clickSelectDropdown(): Promise<void> {
+    await this.clickableParentLocator.click({ position: { x: 5, y: 5 } });
```
Comment (djanicekpach) [inline webui/react/src/e2e/models/hew/DataGrid.ts:302]: Is this OK on smaller screens? I know we haven't done much with it, but it'd be good to think about.

File: webui/react/src/e2e/models/hew/DataGrid.ts [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+  /**
+   * Returns the index of the row. Start counting at 0.
+   */
+  async getIndex(): Promise<number> {
+    const value = await this.pwLocator.getAttribute(this.indexAttribute);
+    if (value === null) {
+      throw new Error(`All rows should have the attribute ${this.indexAttribute}`);
+    }
+    return +value - 2;
```
Comment (djanicekpach) [inline webui/react/src/e2e/models/hew/DataGrid.ts:193]: maybe check that value is a legit number here before casting with +. This seems like the type of thing that could cast an unexpected value without error.

File: webui/react/src/e2e/models/hew/DataGrid.ts [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+    );
+  }
+
+  /**
+   * Returns a list of rows that match the condition provided
+   * @param {(row: RowType) => Promise<boolean>} condition - function which tests each row against a condition
+   */
+  async filterRows(condition: (row: RowType) => Promise<boolean>): Promise<RowType[]> {
+    return (
+      await Promise.all(
```
Comment (djanicekpach) [inline webui/react/src/e2e/models/hew/DataGrid.ts:128]: This data grid stuff does not seem thread-safe. Not an issue now, but it probably will be later. If another test is creating/deleting experiments any index we are tracking goes out the window, and I think these multi-promise loops could wind up with contradictory data. If we have to use the index to interact with this, maybe we have a semaphore around experiments? I'm open to ideas hee too. I'm thinking it would be nice to have separate read/write locks. i.e.

- `Lock/UnlockWrite()`: prevents reading from experiment list and writing experiments. Test code writing to experiments would be expected to acquire this lock before making changes.
- `Lock/UnlockRead()`: prevents writing to experiments, but allows other tests to read. Tests reading experiment lists(like this one) would be expected to acquire this lock before reading.  

In other words,`LockWrite` would block `LockRead` and additional `LockWrite`s. `LockRead` would block `LockWrite`, but not additional `LockRead`s

**Review by djanicekpach — CHANGES_REQUESTED** [review]

File: webui/react/src/e2e/models/BasePage.ts [hunk 6]
```diff
@@ -25,14 +29,22 @@ export abstract class BasePage {
...
-    await this._page.goto(this.url);
-    if (waitForURL) {
+  async goto(
+    { url, verify = true }: { url: string | RegExp; verify?: boolean } = this,
+  ): Promise<BasePage> {
+    if (url instanceof RegExp) {
+      throw new Error(`${typeof this}.url is a regular expression. Please provide a url to visit.`);
+    }
+    await this._page.goto(url);
+    if (verify) {
```
[the remark this answers is not in the document]
Comment (JComins000) [reply webui/react/src/e2e/models/BasePage.ts:45]: true. regex can be quite malleable so i'd like to leave it like this for now. maybe the default option can be specific to `verify = (typeof URL === "string")` or something

File: webui/react/src/e2e/models/components/F_ExperiementList.ts [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+import { DataGrid, HeadRow, Row } from 'e2e/models/hew/DataGrid';
+import { Pagination } from 'e2e/models/hew/Pagination';
+
+/**
+ * Returns a representation of the F_ExperiementList component.
+ * This constructor represents the contents in src/components/F_ExperiementList.tsx.
+ * @param {object} obj
+ * @param {CanBeParent} obj.parent - The parent used to locate this F_ExperiementList
+ */
+export class F_ExperiementList extends BaseReactFragment {
```
[the remark this answers is not in the document]
Comment (JComins000) [reply webui/react/src/e2e/models/components/F_ExperiementList.ts:14]: yes it collides. it's version 2 and it's named the same way in the source

File: webui/react/src/e2e/models/components/F_ExperiementList.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  readonly pagination: Pagination = new Pagination({ parent: this });
+}
+
+class ExperimentHeadRow extends HeadRow {}
+class ExperimentRow extends Row<ExperimentRow, ExperimentHeadRow> {
+  /**
+   * Clicks the row's id cell
+   */
+  async clickID() {
+    await this.clickX(50);
```
[the remark this answers is not in the document]
Comment (JComins000) [reply webui/react/src/e2e/models/components/F_ExperiementList.ts:39]: yup i'll make a constant

File: webui/react/src/e2e/models/hew/DataGrid.ts (hunk 5 above)
Comment (JComins000) [reply webui/react/src/e2e/models/hew/DataGrid.ts:128]: I think I would need to refresh the page before seeing any new experiments, right?

File: webui/react/src/e2e/models/hew/DataGrid.ts (hunk 4 above)
Comment (JComins000) [reply webui/react/src/e2e/models/hew/DataGrid.ts:193]: I can add it to line 190

File: webui/react/src/e2e/models/hew/DataGrid.ts (hunk 3 above)
Comment (JComins000) [reply webui/react/src/e2e/models/hew/DataGrid.ts:302]: ehh let's deal with that when we get to it

File: webui/react/src/e2e/models/hew/DataGrid.ts (hunk 5 above)
Comment (ashtonG) [reply webui/react/src/e2e/models/hew/DataGrid.ts:129]: no -- the experiment list requests the current page every 10 seconds

Comment (JComins000) [reply webui/react/src/e2e/models/hew/DataGrid.ts:129]: useful! I think in that case the playwright locator would expire and throw an error. Maybe I could handle that error somehow. I left a todo since it's not in the scope of this ticket

**Review by djanicekpach — APPROVED** [review]

**Review by ashtonG — APPROVED** [review]

**Review by keita-determined — APPROVED** [review]
